django/api/views.py

```diff
@@ -1,21 +1,51 @@
-from rest_framework import generics
+from rest_framework import generics, mixins, status
 from rest_framework.permissions import IsAuthenticated
-from rest_framework import mixins
+from rest_framework.response import Response
+from rest_framework.settings import api_settings
 from rest_framework.viewsets import GenericViewSet
 from rest_framework_gis.filters import InBBoxFilter
 from django.contrib.auth.models import User
 from api.models import Restaurant
 from api.serializers import UserSerializer, RestaurantSerializer
+from api.permissions import IsAuthenticatedOrCreate
 
 
-class UserViewSet(mixins.RetrieveModelMixin, GenericViewSet):
+class UserList(generics.ListAPIView):
     permission_classes = (IsAuthenticated,)
     queryset = User.objects.all()
     serializer_class = UserSerializer
 
+
+class UserViewSet(mixins.RetrieveModelMixin, mixins.UpdateModelMixin, GenericViewSet):
+    permission_classes = (IsAuthenticatedOrCreate,)
+    model_class = User
+
     def get_object(self):
         return self.request.user
 
+    def get_serializer_class(self):
+        if self.request.data.__contains__("restaurant_name"):
```

Comment:
Also no change necessary here, but as I'm looking at this and some of the hoops that DRF is making you jump through in order for it to work, I'm wondering whether making the switch on the frontend might eliminate some of them. You'd have to define two endpoints, and it's definitely not as RESTy, because you'd have something like `/api/User/create-with-restaurant/`, but making the two pieces of functionality very, very separate might allow you to have two simple logic paths to follow on the backend rather than one more complex one.